### PR TITLE
refactor(connector): [Shift4]Enhance currency Mapping with ConnectorCurrencyCommon Trait

### LINKED ISSUE #2245: [FEATURE]: [Shift4] Currency Unit Conversion
### :memo: Feature Description

- Each currency can be described in terms of base or minor units. 
- For instance, when dealing with USD currency, the Cent is considered as the minor unit, while the Dollar is considered the base unit. 
- In Hyperswitch, the amount value is expected to be always provided in minor units. 
- For example: In USD, If the amount is 1 in base unit (Dollar), then it will be equivalent to 100 in minor units (Cent)
- Some of the connectors integrated require the amount to be converted to a desirable unit before being passed to them. 
- We have functions `to_currency_base_unit` and `to_currency_lower_unit` in place to convert the minor unit amount to its decimal equivalent and vice versa. These conversions are handled explicitly based on the connector.

### :hammer: Possible Implementation

- ConnectorCommon trait have been implemented for the connector. 
- This trait contains `get_currency_unit` method. This method needs to be implemented. 
- It will define what type of conversion needs to be done during `connector_router_data` creation. Concurrently handle the creation of `connector_router_data` to be passed for the connector request body creation.
- You can check this PR for further reference https://github.com/juspay/hyperswitch/pull/2196

:bookmark: Note: All the changes needed should be contained within `hyperswitch/crates/router/src/connector/`

### :package: Have you spent some time checking if this feature request has been raised before?

- [X] I checked and didn't find a similar issue

### :package: Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by sanketmp**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This pull request introduces the get_currecny_unit from ConnectorCommon trait for shift4. This function allows connectors to declare their accepted currency unit as either "Base" or "Minor".

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Closes #2245 

## How did you test it?

(Please review the changes and let me know if any further issues or changes to be made.)

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by srujanchikke** [comment]
Hi @sanketmp ,

Please use following cargo commands -

`cargo clippy --all-features`
ref -[ Cargo commands](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md#cargo-commands), and fix the errors thrown by clippy for the checks to pass and also use `cargo +nightly fmt` to fix the failing formatting check

**Comment by sanketmp** [comment]
ran the `cargo +nightly fmt` and gave no errors. Please review @srujanchikke

**Comment by srujanchikke** [comment]
> cargo clippy --all-features

You might also should run `cargo clippy --all-features`

**Comment by sanketmp** [comment]
> > cargo clippy --all-features
> 
> You might also should run `cargo clippy --all-features`

I have done that too. I don't know what is happening.

File: crates/router/src/connector/shift4/transformers.rs [hunk 1]
```diff
@@ -120,17 +151,27 @@ pub enum CardPayment {
...
+impl<T>
+    TryFrom<
+        &Shift4RouterData<
+            &types::RouterData<T, types::PaymentsAuthorizeRouterData, types::PaymentsResponseData>,
+        >,
+    > for Shift4PaymentsRequest
 {
-    type Error = Error;
+    type Error = error_stack::Report<errors::ConnectorError>;
+    // fn try_from(
```
Comment (prasunna09) [inline crates/router/src/connector/shift4/transformers.rs:162]: can you please remove these comments?

File: crates/router/src/connector/shift4/transformers.rs [hunk 2]
```diff
@@ -120,17 +151,27 @@ pub enum CardPayment {
...
+    //     item: &types::RouterData<T, types::PaymentsAuthorizeData, types::PaymentsResponseData>,
+    // ) -> Result<Self, Self::Error> {
     fn try_from(
-        item: &types::RouterData<T, types::PaymentsAuthorizeData, types::PaymentsResponseData>,
+        value: (&Shift4RouterData<
```
Comment (prasunna09) [inline crates/router/src/connector/shift4/transformers.rs:166]: can you please rename value to item
```suggestion
        item: (&Shift4RouterData<
```

File: crates/router/src/connector/shift4/transformers.rs [hunk 3]
```diff
@@ -120,17 +151,27 @@ pub enum CardPayment {
...
+            &types::PaymentsAuthorizeRouterData,
+            types::PaymentsResponseData,
+        >),
     ) -> Result<Self, Self::Error> {
-        let submit_for_settlement = item.request.is_auto_capture()?;
-        let amount = item.request.amount.to_string();
-        let currency = item.request.currency;
-        let payment_method = Shift4PaymentMethod::try_from(item)?;
+        let submit_for_settlement = value.request.is_auto_capture()?;
+        let amount = value.request.amount.to_string();
```
Comment (prasunna09) [inline crates/router/src/connector/shift4/transformers.rs:172]:
```suggestion
        let amount = item.amount.clone();
```

**Review by prasunna09 — CHANGES_REQUESTED** [review]

**Comment by sanketmp** [comment]
Hello @prasunna09, made the requested changes.

File: crates/router/src/connector/shift4.rs [hunk 4]
```diff
@@ -61,6 +61,10 @@ impl ConnectorCommon for Shift4 {
         "shift4"
     }
 
+    fn get_currency_unit(&self) -> api::CurrencyUnit {
+        api::CurrencyUnit::Base
```
[the remark this answers is not in the document]
Comment (sanketmp) [reply crates/router/src/connector/shift4.rs]: made the changes

**Comment by SanchithHegde** [comment]
Hey @sanketmp, could you address the failing CI checks?

**Comment by sanketmp** [comment]
I couldn't figure out @SanchithHegde

**Comment by SanchithHegde** [comment]
@sanketmp As mentioned earlier in https://github.com/juspay/hyperswitch/pull/2557#issuecomment-1759244561, you must run `cargo clippy --all-features` and address the errors thrown.

**Comment by sanketmp** [comment]
I have done that but I'll try again

**Pull request closed without merging**